core/src/avm2/globals/flash/geom/Vector3D.as

```diff
@@ -0,0 +1,146 @@
+        }
+
+        public function get length():Number {
+            return Math.sqrt(this.lengthSquared);
+        }
+
+        public function toString():String {
+            return "Vector3D(" + this.x + ", " + this.y + ", " + this.z + ")";
+        }
+
+        public function add(a:Vector3D):Vector3D {
+            return new Vector3D(this.x + a.x, this.y + a.y, this.z + a.z);
+        }
+
+        public function subtract(a:Vector3D):Vector3D {
+            return new Vector3D(this.x - a.x, this.y - a.y, this.z - a.z);
+        }
+
+        public function incrementBy(a:Vector3D):void {
+            this.x += a.x;
+            this.y += a.y;
+            this.z += a.z;
+        }
+
+        public function decrementBy(a:Vector3D):void {
+            this.x -= a.x;
+            this.y -= a.y;
+            this.z -= a.z;
+        }
+
+        public function clone():Vector3D {
+            return new Vector3D(this.x, this.y, this.z, this.w);
+        }
+
+        public function copyFrom(sourceVector3D:Vector3D):void {
+            this.x = sourceVector3D.x;
+            this.y = sourceVector3D.y;
+            this.z = sourceVector3D.z;
+            // w is not touched!
+        }
+
+        public function equals(toCompare:Vector3D, allFour:Boolean = false):Boolean {
+            return this.x == toCompare.x && this.y == toCompare.y && this.z == toCompare.z && (!allFour || this.w == toCompare.w);
+        }
+
+        public function nearEquals(toCompare:Vector3D, tolerance:Number, allFour:Boolean = false):Boolean {
+            if (allFour)
+                // TODO seems like in this case, tolerance is just ignored???
+                return this.equals(toCompare, true);
+
+            return (Math.abs(this.x - toCompare.x) < tolerance)
+                && (Math.abs(this.y - toCompare.y) < tolerance)
+                && (Math.abs(this.z - toCompare.z) < tolerance);
+        }
+
+        public function setTo(xa:Number, ya:Number, za: Number):void {
+            this.x = xa;
+            this.y = ya;
+            this.z = za;
+        }
+
+        public function scaleBy(s:Number):void {
+            this.x *= s;
+            this.y *= s;
+            this.z *= s;
+        }
+
+        public function negate():void {
+            this.x *= -1;
+            this.y *= -1;
+            this.z *= -1;
+        }
+
+        public function project():void {
+            this.x /= this.w;
+            this.y /= this.w;
+            this.z /= this.w;
+            // w is unchanged!
+        }
+
+        public function normalize():Number {
+            var len : Number = this.length;
+
+            if (len == 0) {
+                this.x = 0;
+                this.y = 0;
+                this.z = 0;
+            }
+            else if (len > 0) {
+                this.x /= len;
+                this.y /= len;
+                this.z /= len;
+            }
+            else {
+                this.x = NaN;
```

Comment:
Unless I miss something, `len` cannot be negative and thus this case is dead code.

Reply:
Right, I thought the same thing, but this is for `undefined`. I copied the test cases from `Point` which had cases with these special values. A comment would probably help, as it really looks strange.

Reply:
I added some comments, if you'd like me to add them to some more places, do tell.